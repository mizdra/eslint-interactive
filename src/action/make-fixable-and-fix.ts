import { access, mkdir, readFile, writeFile } from 'node:fs/promises';
import { dirname } from 'node:path';
import type { Remote } from 'comlink';
import type { ESLint } from 'eslint';
import { promptToInputReuseScript } from '../cli/prompt.js';
import { fixingSpinner } from '../cli/spinner.js';
import type { Undo } from '../core.js';
import type { SerializableCore } from '../core-worker.js';
import {
  editFileWithEditor,
  generateExampleFixableMakerScriptFilePath,
  generateFixableMakerScriptFilePath,
} from '../util/filter-script.js';

export async function doMakeFixableAndFixAction(
  core: Remote<SerializableCore>,
  results: ESLint.LintResult[],
  selectedRuleIds: string[],
): Promise<Undo> {
  const exampleScript = await readFile(generateExampleFixableMakerScriptFilePath(), 'utf8');
  const fixableMakerScriptFilePath = generateFixableMakerScriptFilePath(selectedRuleIds);
  const isFixableMakerScriptExist = await access(fixableMakerScriptFilePath)
    .then(() => true)
    .catch(() => false);
  if (isFixableMakerScriptExist) {
    const reuseScript = await promptToInputReuseScript();
    if (!reuseScript) {
      await writeFile(fixableMakerScriptFilePath, exampleScript);
    }
  } else {
    await mkdir(dirname(fixableMakerScriptFilePath), { recursive: true }); // Create the directory because it might not exist
    await writeFile(fixableMakerScriptFilePath, exampleScript);
  }
  console.log('Opening editor...');

  const fixableMakerScript = await editFileWithEditor(fixableMakerScriptFilePath);
  const undo = await fixingSpinner(async () => core.makeFixableAndFix(results, selectedRuleIds, fixableMakerScript));
  return undo;
}
