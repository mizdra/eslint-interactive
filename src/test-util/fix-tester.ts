import type { Linter, Rule } from 'eslint';
import { verifyAndFix } from '../eslint/linter.js';
import { LegacyESLint } from '../eslint/use-at-your-own-risk.js';
import type { FixContext } from '../fix/index.js';
import { plugin } from '../plugin.js';

const DEFAULT_FILENAME = 'test.js';

/**
 * The type representing the test case.
 */
type TestCase<FixArgs> = {
  /**
   * The filename of the code.
   */
  filename?: string;
  /**
   * The code to be fixed.
   */
  code: string | string[];
  /**
   * The rules to fix.
   */
  rules: Partial<Linter.RulesRecord>;
  /**
   * The arguments to pass to the fix function.
   */
  args?: FixArgs;
};

type TestResult = string | null;

/**
 * The test utility for the fix.
 */
export class FixTester<FixArgs> {
  private fixCreator: (context: FixContext, args: FixArgs) => Rule.Fix[];
  private defaultFixArgs: FixArgs;
  private defaultLinterConfig: Linter.LegacyConfig;
  constructor(
    fixCreator: (context: FixContext, args: FixArgs) => Rule.Fix[],
    defaultFixArgs: FixArgs,
    defaultLinterConfig: Linter.LegacyConfig,
  ) {
    this.fixCreator = fixCreator;
    this.defaultFixArgs = defaultFixArgs;
    this.defaultLinterConfig = defaultLinterConfig;
  }
  /**
   * Test the fix.
   * @param testCase The test case.
   * @returns The fixed code. If the fix skipped, null is returned.
   */
  async test(testCase: TestCase<FixArgs>): Promise<TestResult> {
    const code = Array.isArray(testCase.code) ? testCase.code.join('\n') : testCase.code;

    const filePath = testCase.filename ?? DEFAULT_FILENAME;

    const ruleIdsToFix = Object.keys(testCase.rules);
    const eslint = new LegacyESLint({
      useEslintrc: false,
      plugins: {
        'eslint-interactive': plugin,
      },
      overrideConfig: {
        ...this.defaultLinterConfig,
        plugins: ['eslint-interactive', ...(this.defaultLinterConfig.plugins ?? [])],
        rules: {
          ...this.defaultLinterConfig.rules,
          ...testCase.rules,
          'eslint-interactive/source-code-snatcher': 'error',
        },
      },
    });
    const fixedResult = await verifyAndFix(eslint, code, filePath, ruleIdsToFix, (context) =>
      this.fixCreator(context, testCase.args ?? this.defaultFixArgs),
    );

    if (fixedResult.fixed) {
      return fixedResult.output;
    } else {
      return null;
    }
  }
}
