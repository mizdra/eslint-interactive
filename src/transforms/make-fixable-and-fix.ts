import { Linter, Rule, SourceCode } from 'eslint';
import { traverse } from 'estraverse';
import type { Node } from 'estree';
import { TransformContext } from '../types.js';
import { unreachable } from '../util/type-check.js';

export type FixableMaker = (message: Linter.LintMessage, node: Node | null) => Rule.Fix | null | undefined;

export type TransformToMakeFixableAndFixArgs = {
  fixableMaker: FixableMaker;
};

/**
 * Check the node is the source of the message.
 */
function isMessageSourceNode(sourceCode: SourceCode, node: Node, message: Linter.LintMessage): boolean {
  if (message.nodeType === undefined) return false;

  // If `nodeType` is exists, `endLine` and `endColumn` must be exists.
  if (message.endLine === undefined || message.endColumn === undefined) return unreachable();
  // If `nodeType` is exists, `range` must be exists.
  if (node.range === undefined) return unreachable();

  const index = sourceCode.getIndexFromLoc({
    line: message.line,
    // NOTE: `column` of `ESLint.LintMessage` is 1-based, but `column` of `ESTree.Position` is 0-based.
    column: message.column - 1,
  });
  const endIndex = sourceCode.getIndexFromLoc({
    line: message.endLine,
    // NOTE: `column` of `ESLint.LintMessage` is 1-based, but `column` of `ESTree.Position` is 0-based.
    column: message.endColumn - 1,
  });
  const nodeType = message.nodeType;

  return node.range[0] === index && node.range[1] === endIndex && node.type === nodeType;
}

function getMessageToSourceNode(sourceCode: SourceCode, messages: Linter.LintMessage[]): Map<Linter.LintMessage, Node> {
  const result = new Map<Linter.LintMessage, Node>();

  traverse(sourceCode.ast, {
    // Required to traverse extension nodes such as `JSXElement`.
    fallback: 'iteration',
    enter(node: Node) {
      for (const message of messages) {
        if (isMessageSourceNode(sourceCode, node, message)) {
          result.set(message, node);
        }
      }
    },
  });
  return result;
}

function generateFixes(context: TransformContext, args: TransformToMakeFixableAndFixArgs): Rule.Fix[] {
  const messageToNode = getMessageToSourceNode(context.sourceCode, context.messages);

  const fixes: Rule.Fix[] = [];
  for (const message of context.messages) {
    const node = messageToNode.get(message) ?? null;
    const fix = args.fixableMaker(message, node);
    if (fix) fixes.push(fix);
  }
  return fixes;
}

/**
 * Create transform to make fixable and fix.
 */
export function createTransformToMakeFixableAndFix(
  context: TransformContext,
  args: TransformToMakeFixableAndFixArgs,
): Rule.Fix[] {
  return generateFixes(context, args);
}
