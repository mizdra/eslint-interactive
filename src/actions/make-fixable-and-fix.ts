import { access, mkdir, readFile, writeFile } from 'fs/promises';
import { dirname } from 'path';
import chalk from 'chalk';
import { ESLint } from 'eslint';
import ora from 'ora';
import { promptToInputReuseScript } from '../cli/prompt.js';
import { Core } from '../core.js';
import { FixableMaker } from '../transforms/make-fixable-and-fix.js';
import {
  editFileWithEditor,
  generateExampleFixableMakerScriptFilePath,
  generateFixableMakerScriptFilePath,
} from '../util/filter-script.js';

export async function doMakeFixableAndFixAction(
  core: Core,
  results: ESLint.LintResult[],
  selectedRuleIds: string[],
): Promise<void> {
  const exampleScript = await readFile(generateExampleFixableMakerScriptFilePath(), 'utf8');
  const fixableMakerScriptFilePath = generateFixableMakerScriptFilePath(selectedRuleIds);
  const isFixableMakerScriptExist = await access(fixableMakerScriptFilePath)
    .then(() => true)
    .catch(() => false);
  if (isFixableMakerScriptExist) {
    const reuseScript = await promptToInputReuseScript();
    if (!reuseScript) {
      await writeFile(fixableMakerScriptFilePath, exampleScript);
    }
  } else {
    // ディレクトリがない可能性を考慮して作成しておく
    await mkdir(dirname(fixableMakerScriptFilePath), { recursive: true });
    await writeFile(fixableMakerScriptFilePath, exampleScript);
  }
  console.log('Opening editor...');

  const fixableMakerScript = await editFileWithEditor(fixableMakerScriptFilePath);
  const fixableMaker = eval(fixableMakerScript) as FixableMaker;
  const fixingSpinner = ora('Making fixable and fixing...').start();
  await core.makeFixableAndFix(results, selectedRuleIds, fixableMaker);
  fixingSpinner.succeed(chalk.bold('Making fixable and fixing was successful.'));
}
