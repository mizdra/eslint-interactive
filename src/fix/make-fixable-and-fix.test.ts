import { basename } from 'node:path';
import { describe, expect, test } from 'vitest';
import { FixTester } from '../test-util/fix-tester.js';
import { createFixToMakeFixableAndFix, FixToMakeFixableAndFixArgs } from './make-fixable-and-fix.js';

const tester = new FixTester<FixToMakeFixableAndFixArgs>(
  createFixToMakeFixableAndFix,
  {
    fixableMaker: (_message, node) => {
      if (!node || !node.range) return null;
      return {
        range: [node.range[0], node.range[0]],
        text: '/* test */',
      };
    },
  },
  { parserOptions: { ecmaVersion: 2020, ecmaFeatures: { jsx: true } } },
);

describe('make-fixable-and-fix', () => {
  test('basic', () => {
    expect(
      tester.test({
        code: 'const a = 1;',
        rules: { 'no-unused-vars': ['error', { varsIgnorePattern: '^_' }] },
        args: {
          fixableMaker: (_message, node) => {
            if (!node || !node.range) return null;
            return { range: [node.range[0], node.range[0]], text: '_' };
          },
        },
      }),
    ).toMatchInlineSnapshot(`"const _a = 1;"`);
  });
  test('can process multiple messages at once', () => {
    expect(
      tester.test({
        code: ['const a = 1;', 'const b = 2;'],
        rules: { 'no-unused-vars': ['error', { varsIgnorePattern: '^_' }] },
        args: {
          fixableMaker: (_message, node) => {
            if (!node || !node.range) return null;
            return { range: [node.range[0], node.range[0]], text: '_' };
          },
        },
      }),
    ).toMatchInlineSnapshot(`
      "const _a = 1;
      const _b = 2;"
    `);
  });
  test('can process messages of multiple rules at once', () => {
    expect(
      tester.test({
        code: ['const a = 1;', 'let b = 2;', 'b++;', 'console.log(b);'],
        rules: { 'no-unused-vars': ['error', { varsIgnorePattern: '^_' }], 'no-plusplus': 'error' },
        args: {
          fixableMaker: (message, node) => {
            if (!node || !node.range) return null;
            if (message.ruleId === 'no-unused-vars') {
              return { range: [node.range[0], node.range[0]], text: '_' };
            } else if (message.ruleId === 'no-plusplus') {
              return { range: [node.range[1] - 2, node.range[1]], text: ' += 1' };
            } else {
              return null;
            }
          },
        },
      }),
    ).toMatchInlineSnapshot(`
      "const _a = 1;
      let b = 2;
      b += 1;
      console.log(b);"
    `);
  });
  test('can process messages on the same line', () => {
    expect(
      tester.test({
        code: ['const a = 1; const b = 2;'],
        rules: { 'no-unused-vars': ['error', { varsIgnorePattern: '^_' }] },
        args: {
          fixableMaker: (_message, node) => {
            if (!node || !node.range) return null;
            return { range: [node.range[0], node.range[0]], text: '_' };
          },
        },
      }),
    ).toMatchInlineSnapshot(`"const _a = 1; const _b = 2;"`);
  });
  test('`fixableMaker` receives the message and node.', () => {
    tester.test({
      code: ['const a = 1;'],
      rules: { 'no-unused-vars': ['error', { varsIgnorePattern: '^_' }] },
      args: {
        fixableMaker: (message, node, context) => {
          expect({
            message,
            node,
            context: {
              ...context,
              // Use only basename, because the path will change depending on the environment.
              filename: basename(context.filename),
              // Take a snapshot of only part of it because `sourceCode` is huge
              sourceCode: { text: context.sourceCode.text },
            },
          }).toMatchSnapshot();
          return null;
        },
      },
    });
  });
  test('node is null if message is not associated with a node', () => {
    tester.test({
      code: ['// this is comment'],
      rules: { 'capitalized-comments': 'error' },
      args: {
        fixableMaker: (_message, node) => {
          expect(node).toBeNull();
          return null;
        },
      },
    });
  });
  test('do not process if `fixableMaker` returns null', () => {
    expect(
      tester.test({
        code: 'const a = 1;',
        rules: { 'no-unused-vars': ['error', { varsIgnorePattern: '^_' }] },
        args: {
          fixableMaker: () => null,
        },
      }),
    ).toMatchInlineSnapshot(`null`);
  });
});
