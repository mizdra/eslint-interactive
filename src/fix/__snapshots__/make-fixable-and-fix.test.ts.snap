// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`make-fixable-and-fix > \`fixableMaker\` receives the message and node. 1`] = `
{
  "context": {
    "filename": "test.js",
    "fixer": {
      "insertTextAfter": [Function],
      "insertTextAfterRange": [Function],
      "insertTextBefore": [Function],
      "insertTextBeforeRange": [Function],
      "remove": [Function],
      "removeRange": [Function],
      "replaceText": [Function],
      "replaceTextRange": [Function],
    },
    "messages": [
      {
        "column": 7,
        "endColumn": 8,
        "endLine": 1,
        "line": 1,
        "message": "'a' is assigned a value but never used. Allowed unused vars must match /^_/u.",
        "messageId": "unusedVar",
        "nodeType": "Identifier",
        "ruleId": "no-unused-vars",
        "severity": 2,
      },
    ],
    "ruleIds": [
      "no-unused-vars",
    ],
    "sourceCode": {
      "text": "const a = 1;",
    },
  },
  "message": {
    "column": 7,
    "endColumn": 8,
    "endLine": 1,
    "line": 1,
    "message": "'a' is assigned a value but never used. Allowed unused vars must match /^_/u.",
    "messageId": "unusedVar",
    "nodeType": "Identifier",
    "ruleId": "no-unused-vars",
    "severity": 2,
  },
  "node": Node {
    "end": 7,
    "loc": SourceLocation {
      "end": Position {
        "column": 7,
        "line": 1,
      },
      "start": Position {
        "column": 6,
        "line": 1,
      },
    },
    "name": "a",
    "parent": Node {
      "end": 11,
      "id": [Circular],
      "init": Node {
        "end": 11,
        "loc": SourceLocation {
          "end": Position {
            "column": 11,
            "line": 1,
          },
          "start": Position {
            "column": 10,
            "line": 1,
          },
        },
        "parent": [Circular],
        "range": [
          10,
          11,
        ],
        "raw": "1",
        "start": 10,
        "type": "Literal",
        "value": 1,
      },
      "loc": SourceLocation {
        "end": Position {
          "column": 11,
          "line": 1,
        },
        "start": Position {
          "column": 6,
          "line": 1,
        },
      },
      "parent": Node {
        "declarations": [
          [Circular],
        ],
        "end": 12,
        "kind": "const",
        "loc": SourceLocation {
          "end": Position {
            "column": 12,
            "line": 1,
          },
          "start": Position {
            "column": 0,
            "line": 1,
          },
        },
        "parent": Node {
          "body": [
            [Circular],
          ],
          "comments": [],
          "end": 12,
          "loc": SourceLocation {
            "end": Position {
              "column": 12,
              "line": 1,
            },
            "start": Position {
              "column": 0,
              "line": 1,
            },
          },
          "parent": null,
          "range": [
            0,
            12,
          ],
          "sourceType": "script",
          "start": 0,
          "tokens": [
            Token {
              "end": 5,
              "loc": SourceLocation {
                "end": Position {
                  "column": 5,
                  "line": 1,
                },
                "start": Position {
                  "column": 0,
                  "line": 1,
                },
              },
              "range": [
                0,
                5,
              ],
              "start": 0,
              "type": "Keyword",
              "value": "const",
            },
            Token {
              "end": 7,
              "loc": SourceLocation {
                "end": Position {
                  "column": 7,
                  "line": 1,
                },
                "start": Position {
                  "column": 6,
                  "line": 1,
                },
              },
              "range": [
                6,
                7,
              ],
              "start": 6,
              "type": "Identifier",
              "value": "a",
            },
            Token {
              "end": 9,
              "loc": SourceLocation {
                "end": Position {
                  "column": 9,
                  "line": 1,
                },
                "start": Position {
                  "column": 8,
                  "line": 1,
                },
              },
              "range": [
                8,
                9,
              ],
              "start": 8,
              "type": "Punctuator",
              "value": "=",
            },
            Token {
              "end": 11,
              "loc": SourceLocation {
                "end": Position {
                  "column": 11,
                  "line": 1,
                },
                "start": Position {
                  "column": 10,
                  "line": 1,
                },
              },
              "range": [
                10,
                11,
              ],
              "start": 10,
              "type": "Numeric",
              "value": "1",
            },
            Token {
              "end": 12,
              "loc": SourceLocation {
                "end": Position {
                  "column": 12,
                  "line": 1,
                },
                "start": Position {
                  "column": 11,
                  "line": 1,
                },
              },
              "range": [
                11,
                12,
              ],
              "start": 11,
              "type": "Punctuator",
              "value": ";",
            },
          ],
          "type": "Program",
        },
        "range": [
          0,
          12,
        ],
        "start": 0,
        "type": "VariableDeclaration",
      },
      "range": [
        6,
        11,
      ],
      "start": 6,
      "type": "VariableDeclarator",
    },
    "range": [
      6,
      7,
    ],
    "start": 6,
    "type": "Identifier",
  },
}
`;
